File: tests/WebSharper.CSharp.Tests/Task.cs
// $begin{copyright}
//
// This file is part of WebSharper
//
// Copyright (c) 2008-2018 IntelliFactory
//
// Licensed under the Apache License, Version 2.0 (the "License"); you
// may not use this file except in compliance with the License.  You may
// obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.  See the License for the specific language governing
// permissions and limitations under the License.
//
// $end{copyright}
using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using WebSharper.Testing;

namespace WebSharper.CSharp.Tests
{
    [JavaScript, Test("C# Task and async tests")]
    public class TaskAsync : TestCategory
    {
        public async Task<int> GetOneAsync()
        {
            var o = Task.FromResult(1);
            return await o;
        }

        [Test("Async/Await")]
        public async Task AsyncAwait()
        {
            var one = await GetOneAsync();
            Equal(one, 1);
        }

        [Test("loop inside Async")]
        public async Task AsyncLoop()
        {
            var i = 0;
            while (i < 10)
            {
                await Task.Delay(0);
                i++;
            }
            Equal(i, 10);
        }

        int Counter = 0;

        async Task IncrementWithCancel(CancellationToken ct)
        {
            for (var i = 0; i < 10; i++)
            {
                await Task.Delay(10);
                ct.ThrowIfCancellationRequested();
                Counter++;
            }
        }

        [Test("Yield")]
        public async Task Yield()
        {
            var one = await GetOneAsync();
            await Task.Yield();
            Equal(one, 1);
        }

        [Test("Task cancellation")]
        public async void TaskCancellation()
        {
            Expect(0);
            var cts = new CancellationTokenSource();
            var task = IncrementWithCancel(cts.Token);
            await Task.Delay(30);
            cts.Cancel();
            await Task.Delay(50);
            IsTrue(task.IsCanceled);
            IsTrue(Counter < 5);
        }

        [Test("Local async function")]
        public async void LocalAsync()
        {
            async Task<int> GetOneAsyncLocal()
            {
                var o = Task.FromResult(1);
                return await o;
            }

            var one = await GetOneAsyncLocal();
            Equal(one, 1);
        }

        [Test("Result")]
        public void Result()
        {
            Equal(Task.FromResult<int>(1).Result, 1);
            Raises(() => Task.FromException<int>(new Exception()).Result);
            Raises(() => new Task<int>(() => 2).Result);
        }

        [Test]
        public async Task WebWorker()
        {
            var worker = new JavaScript.Worker(self =>
            {
                self.Onmessage = e =>
                {
                    self.PostMessage(GlobalClass.GlobalFunction((string)e.Data));
                };
            });
            var t = new TaskCompletionSource<string>();
            worker.Onmessage = e =>
            {
                t.SetResult("The worker replied: " + (string)e.Data);
            };
            worker.PostMessage("Hello world!");
            var res = await t.Task;
            worker.Terminate();
            Equal(res, "The worker replied: [worker] Hello world!");
        }
    }

    // This needs to be a separate class, otherwise the web worker bundle
    // will include QUnit as a dependency.
    [JavaScript]
    public static class GlobalClass
    {
        public static string GlobalFunction(string s)
        {
            return "[worker] " + s;
        }

    }
}

File: tests/WebSharper.Tests/Task.fs
// $begin{copyright}
//
// This file is part of WebSharper
//
// Copyright (c) 2008-2018 IntelliFactory
//
// Licensed under the Apache License, Version 2.0 (the "License"); you
// may not use this file except in compliance with the License.  You may
// obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.  See the License for the specific language governing
// permissions and limitations under the License.
//
// $end{copyright}

module WebSharper.Tests.Task

open System.Threading.Tasks

open WebSharper
open WebSharper.Testing

[<JavaScript>]
let Tests =
    TestCategory "Task" {
        Test "Run" {
            let taskDone = ref false  
            let task = Task.Run(fun () -> taskDone := true)
            do! Async.AwaitTask task
            isTrue task.IsCompleted
            isFalse task.IsCanceled
            isFalse task.IsFaulted
            isTrue !taskDone

            let task2 = Task.Run(fun () -> 1 + 2)
            let! task2res = Async.AwaitTask task2
            equal task2res 3
        }

        Test "Failed" {
            let task = Task.Run(fun () -> failwith<unit> "error")
            do! async { 
                try do! Async.AwaitTask task
                with _ -> ()
            }
            isTrue task.IsCompleted
            isTrue task.IsFaulted
            isFalse task.IsCanceled
            equal task.Exception.InnerExceptions.[0].Message "error"
        }

        Test "StartAsTask" {
            let task = Async.StartAsTask (async {return 2 })
            let! taskRes = Async.AwaitTask task
            equal taskRes 2

            let task2 = Async.StartAsTask (async { failwith "error" })
            let! task2Error =
                async {
                    try
                        let! _ = Async.AwaitTask task2
                        return Some "awaiting failing task should fail"
                    with 
                    | (:? System.AggregateException as e) ->                        
                        if (e).InnerExceptions.[0].Message = "error" then
                            return None
                        else return Some "wrong error propagated"
                    | _ -> return Some "error is not AggregateException"
                }
            equal task2Error None    
        }

        Test "ContinueWith" {
            let state = ref ""
            let task = new Task(fun () -> state := "A")
            equal !state ""
            equal task.Status TaskStatus.Created
            let task2 = task.ContinueWith(fun t -> if not t.IsFaulted then state := !state + "B")
            equal task2.Status TaskStatus.WaitingForActivation
            task.Start()
            do! Async.AwaitTask task2
            isTrue task2.IsCompleted
            equal !state "AB"
        }

        Test "Task computation expression" {
            let task1 = task { return 1 }
            let! task1Res = Async.AwaitTask task1
            equal task1Res 1
            let task2 = task { return! task1 }
            let! task2Res = Async.AwaitTask task2
            equal task2Res 1
            let task3 = task { return 2 }
            let task4 =
                task {
                    let! r1 = task1
                    let! r3 = task3
                    return r1 + r3
                }
            let! task4Res = Async.AwaitTask task4
            equal task4Res 3
        }

        Test "For" {
            let x = task {
                let l = ref []
                for i in 1 .. 3 do 
                    l := i :: !l
                return  !l
            }
            let! xRes = Async.AwaitTask x
            equal xRes [ 3; 2; 1 ]
        }

        Test "TryWith" {
            let ops = ref ""
            let x = task {
                try
                    ops := !ops + "A"
                    failwith "error"
                    ops := !ops + "A"
                with _ ->
                    ops := !ops + "B"
                return !ops
            }
            let! xRes = Async.AwaitTask x
            equal xRes "AB"
        }

        Test "TryFinally" {
            let ops = ref ""
            let x = task {
                try
                    try
                        ops := !ops + "A"
                        failwith "error"
                        ops := !ops + "A"
                    finally ops := !ops + "F"
                with _ -> ()
                return !ops
            }
            let! xRes = Async.AwaitTask x
            equal xRes "AF"
        }
    }
